tests/themis/themis_secure_cell.c

```diff
@@ -32,6 +33,94 @@ static char passwd[] = "password";
 static char message[] = "secure cell test message by Mnatsakanov Andrey from Cossack Labs";
 static char user_context[] = "secure cell user context";
 
+static bool non_zero_buffer(const uint8_t* buffer, size_t length)
+{
+    for (size_t i = 0; i < length; i++) {
+        if (buffer[i] != 0) {
+            return true;
+        }
+    }
+    return false;
+}
+
+static bool different_buffers(const uint8_t* buffer1, size_t length1, const uint8_t* buffer2, size_t length2)
+{
+    size_t min = (length1 < length2) ? length1 : length2;
+    if (memcmp(buffer1, buffer2, min) != 0) {
+        return true;
+    }
+    return length1 != length2;
+}
+
+static void secure_cell_key_generation(void)
+{
+    themis_status_t res;
+    uint8_t master_key_1[MAX_KEY_SIZE];
+    uint8_t master_key_2[MAX_KEY_SIZE];
+    size_t master_key_1_length = 0;
+    size_t master_key_2_length = 0;
+
+    /*
+     * Normal usage of key generation API.
+     */
+
+    res = themis_gen_sym_key(NULL, &master_key_1_length);
+    testsuite_fail_unless(res == THEMIS_BUFFER_TOO_SMALL, "keygen: query key size");
+    testsuite_fail_unless(master_key_1_length > 0, "keygen: recommends non-empty key");
```

Comment:
maybe will be better to check that result length is `THEMIS_SYM_KEY_LENGTH`. with this check we will be sure that our generation function use this definition, not some hardcoded values or took elsewhere.

Reply:
This constant is not exported by design so that the users won't think that keys always have some fixed length, won't start preallocating fixed buffers, etc. We don't export key length constants for asymmetric keys too. The users just need to know that key length is some positive number. Currently we allow anything other than zero.

This check ensures that we never return zero length as it fails both key generation (in its current implementation) and future use of invalid key with Secure Cell (which does not allow zero-sized keys). I don't consider this a particular magic number.

If anything, I'd probably check that it's `>= 32` — so that we never shorten the keys, only expand them  — but that will be a magic number for real. If we export and use a constant directly then this check will allow us to redefine it to 16 without breaking tests.

Thus... I'd leave it as is.

Reply:
we don't export but we should be sure that this `define` used there and works as expected. it is not integration tests, it is unit test that verifies that function works as expected. we as authors expect that the key should be generated with predefined default value and check that it is expected value imho

Reply:
We don't really have a suitable header to put it into, like <themis/secure_keygen_t.h> and I don't really want to introduce a new one just for this constant. Putting it into some random private header like <themis/secure_session_t.h> is also not nice.

I'd like to have this constant not visible when including <themis/themis.h> so that we don't have to maintain it as a part of public API. If it's visible then someone will notice it and use it, they could very well put a `uint8_t themis_key[THEMIS_SYM_KEY_LENGTH]` somewhere into their data structures, then write them to disk, and now we cannot ever change this length to 64 because this breaks binary compatibility for the users.

If anything, I'd probably go with a copy for tests:

```
#define DEFAULT_SYM_KEY_LENGTH 32
```

then check that we return _this value_ when the buffer is too small. It's not the same macro constant, but it has the same value, so if anyone changes THEMIS_SYM_KEY_LENGTH then they'll have to change it in the other place as well to fix the test. It's a bit more strict than checking for `> 0` and avoids unnecessary exports.

Any second opinion? @vixentael, @ignatk, @storojs72?

Reply:
anyway, we will describe this in our docs, that default behavior of our function for key generation is to use random key with length=32. imho it's not some private value. it's the same as we describe that we use aes with [256 block size](https://docs.cossacklabs.com/pages/documentation-themis/#selecting-the-default-algorithm-parameters). 
and yes, more opinion will be great to choose the final decision)